### PR TITLE
docs(applications): add documentation for healthchecks

### PR DESCRIPTION
**Pull request by bacongobbler**
This is a summarized write-up of the proposal made in deis/controller#603. Note that I'm still a bit iffy on how the client should handle the [HTTP header section](http://kubernetes.io/docs/user-guide/liveness/) in the user guide, but it's up for pondering!

### CODE REVIEW

File: src/applications/managing-app-configuration.md [hunk 1]
```diff
@@ -54,68 +54,106 @@ Detachments can be performed with `deis config:unset`.
...
-
-**HEALTHCHECK\_INITIAL_DELAY**
-
-Number of seconds before health checks starts checking the application after the start of the pod.
-
-This is useful to set if the application takes a while to get setup, due to migrations, cache warming or otherwise. Prevents Kubernetes (and Deis Workflow) from replacing an application pod in its setup phase.
-
-Defaults to 50 seconds
+```
+$ deis healthchecks:set liveness httpGet / 80 --initial-delay-timeout 30 --timeout-seconds 1
```
Comment (kmala) [inline src/applications/managing-app-configuration.md]: i think path should be optional same as kubernetes keeps it and defaults to /

Comment (bacongobbler) [inline src/applications/managing-app-configuration.md]: so like `deis healthchecks:set liveness httpGet 80`, but also allow an optional argument to specify a path via `deis healthchecks:set liveness httpGet 80 /index.html`?

Comment (kmala) [inline src/applications/managing-app-configuration.md]: like `deis healthchecks:set liveness httpGet 80 --path /index.html` how we are treating other optional parameters

Comment (bacongobbler) [inline src/applications/managing-app-configuration.md]: makes sense. I'll make that change.

File: src/applications/managing-app-configuration.md [hunk 2]
```diff
@@ -54,68 +54,122 @@ Detachments can be performed with `deis config:unset`.
...
+`tcpSocket` depending on the type of probe the container requires.
 
-Available configuration options for health checks are the following, including defaults:
+A liveness probe is useful for applications running for long periods of time, eventually
+transitioning to broken states and cannot recover except by restarting them.
 
-**HEALTHCHECK_URL**
+Other times, a readiness probe is useful when the container is only temporarily unable to serve,
+and will recover on its own. In this case, if a container fails its readiness probe, the container
+will not be shut down, but rather the container will stop receiving incoming requests.
```
Comment (helgi) [inline src/applications/managing-app-configuration.md]: Would be good to also make the distinction that `readiness` is used more so during `boot time` and `liveness` is a day to day checks (I'm wording this poorly but I hope you get the gist)

File: src/applications/managing-app-configuration.md [hunk 3]
```diff
@@ -54,68 +54,122 @@ Detachments can be performed with `deis config:unset`.
...
 
-Path in the application to use for health check, such as /healthz - Query Parameters are not supported.
+`httpGet` probes are just as it sounds: it performs a HTTP GET operation on the Container. A
+response code inside the 200-399 range is considered a pass.
 
-This value needs to be set for any health check to happen. Needs to accept a HTTP GET request and return a HTTP status between 200-399.
+`exec` probes run a command inside the Container to determine its health. An exit code of zero is
+considered a pass, while a non-zero status code is considered a fail.
```
Comment (helgi) [inline src/applications/managing-app-configuration.md]: May want to include that the user can run a command such as `env` or to run a script they may have in their application. I know running a command implies both but I want to be very clear what is possible

**Comment by helgi** [comment]
Minor suggestions but otherwise LGTM

File: src/applications/managing-app-configuration.md (hunk 2 above)
Comment (bacongobbler) [inline src/applications/managing-app-configuration.md]: I got what you're saying. I ripped this straight off the kubernetes docs but I can add more examples on what is more applicable in certain situations

File: src/applications/managing-app-configuration.md (hunk 3 above)
Comment (bacongobbler) [inline src/applications/managing-app-configuration.md]: How would you suggest we expand on this? Can't think of anything off the top of my head on which we could make "run a command inside the container" more clearer than it already is :)

Comment (helgi) [inline src/applications/managing-app-configuration.md]: > `exec` probes run a command inside the Container to determine its health, such as running `ps aux | grep <my process>` or run a script within your application that determines when the application is ready

Or something like that

**Comment by kmala** [comment]
@bacongobbler I think it would be better to add some documentation about the current support for the previous configuration using the config:set

**Comment by bacongobbler** [comment]
@kmala I'd prefer to not document legacy code. Existing users can still use it, but we shouldn't point out *how* to use it if we're trying to get away from that.

File: src/applications/managing-app-configuration.md (hunk 2 above)
Comment (bacongobbler) [inline src/applications/managing-app-configuration.md]: added "A liveness probe is useful for applications running for long periods of time". That should make it more clear.

**Pull request merged**